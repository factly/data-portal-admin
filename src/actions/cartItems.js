import axios from '../utils/axios';
import {
  ADD_CARTITEMS,
  SET_CARTITEM_LOADING,
  SET_CARTITEM_REQUEST,
  SET_CARTITEM_IDS,
  CARTITEM_API,
} from '../constants/cartItems';
import { addProducts } from './products';
import { getIds, getValues, deleteKeys, buildObjectOfItems } from '../utils/objects';

export const loadCartItems = (id, page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      cartItems: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCartItemIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CARTITEM_API(id)}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { id, page: page, limit: limit, ids: currentPageIds };
    dispatch(setCartItemRequest(currentReq, total));
    dispatch(addCartItems(nodes));
    dispatch(setCartItemIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CARTITEM_LOADING,
    payload: { loading },
  };
};

export const addCartItems = (cartItems) => (dispatch) => {
  const products = getValues(cartItems, 'product');
  dispatch(addProducts(products));

  dispatch({
    type: ADD_CARTITEMS,
    payload: {
      cartItems: buildObjectOfItems(deleteKeys(cartItems, ['product'])),
    },
  });
};

const setCartItemRequest = (req, total) => {
  return {
    type: SET_CARTITEM_REQUEST,
    payload: { req, total },
  };
};

const setCartItemIds = (ids) => {
  return {
    type: SET_CARTITEM_IDS,
    payload: { ids },
  };
};
