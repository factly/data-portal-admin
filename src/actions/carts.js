import axios from '../utils/axios';
import {
  baseUrl,
  LOADING_CARTS,
  LOADING_CART_DETAILS,
  LOAD_CARTS_SUCCESS,
  LOAD_CARTS_FAILURE,
  GET_CART_ITEMS_SUCCESS,
  GET_CART_ITEMS_FAILURE,
} from '../constants/carts';

export const loadCarts = (page, limit) => {
  return async (dispatch, getState) => {
    let url = baseUrl;
    if (page && limit) {
      url = `${url}?page=${page}&limit=${limit}`;
    }

    dispatch(loadingCarts());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(loadCartsFailure(error.message));
    });

    if (response) {
      dispatch(loadCartsSuccess(response.data));
    }
  };
};

export const getCartItems = (id, page, limit) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}/items`;
    if (page && limit) {
      url = `${url}?page=${page}&limit=${limit}`;
    }

    dispatch(loadingCartDetails());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(getCartItemsFailure(error.message));
    });

    if (response) {
      dispatch(getCartItemsSuccess(response.data));
    }
  };
};

const loadingCarts = () => {
  return {
    type: LOADING_CARTS,
  };
};

const loadingCartDetails = () => {
  return {
    type: LOADING_CART_DETAILS,
  };
};

const loadCartsSuccess = (data) => {
  return {
    type: LOAD_CARTS_SUCCESS,
    payload: {
      items: data.nodes,
      total: data.total,
    },
  };
};

const loadCartsFailure = (message) => {
  return {
    type: LOAD_CARTS_FAILURE,
    payload: message,
  };
};

const getCartItemsSuccess = (data) => {
  return {
    type: GET_CART_ITEMS_SUCCESS,
    payload: {
      items: data.nodes,
      total: data.total,
    },
  };
};

const getCartItemsFailure = (message) => {
  return {
    type: GET_CART_ITEMS_FAILURE,
    payload: message,
  };
};
