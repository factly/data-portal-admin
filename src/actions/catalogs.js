import axios from '../utils/axios';
import {
  baseUrl,
  ADD_CATALOGS_LIST_REQUEST,
  SET_CATALOGS_LIST_CURRENT_PAGE,
  GET_CATALOG_SUCCESS,
  GET_CATALOG_FAILURE,
  LOADING_CATALOGS,
  LOAD_CATALOGS_SUCCESS,
  SET_CATALOGS_LIST_TOTAL,
  LOAD_CATALOGS_FAILURE,
  CREATING_CATALOG,
  CREATE_CATALOG_SUCCESS,
  CREATE_CATALOG_FAILURE,
  UPDATING_CATALOG,
  UPDATE_CATALOG_SUCCESS,
  UPDATE_CATALOG_FAILURE,
  DELETING_CATALOG,
  DELETE_CATALOG_SUCCESS,
  DELETE_CATALOG_FAILURE,
} from '../constants/catalogs';
import { loadProductsSuccess } from './products';
import { getIds, buildObjectOfItems, getValues } from '../utils/objects';

export const loadCatalogs = (page, limit) => {
  return async (dispatch, getState) => {
    let url = baseUrl;
    if (page && limit) {
      url = `${url}?page=${page}&limit=${limit}`;
    }

    const {
      catalogs: { req },
    } = getState();

    let found = false;
    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
        found = true;
      }
    }

    if (found) {
      dispatch(setListCurrentPage(ids));
      return;
    }

    dispatch(loadingCatalogs());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(loadCatalogsFailure(error.message));
    });

    if (response) {
      const { nodes, total } = response.data;
      const currentPageIds = getIds(nodes);
      const req = { page: page, limit: limit, ids: currentPageIds };

      const products = getValues(nodes, 'products').filter((product) => product);
      products && dispatch(loadProductsSuccess(products));

      nodes.forEach((catalog) => {
        catalog.products = getIds(catalog.products || []);
      });

      dispatch(addListRequest(req));
      dispatch(loadCatalogsSuccess(nodes));
      dispatch(setListCurrentPage(currentPageIds));
      dispatch(setCatalogsListTotal(total));
    }
  };
};

export const createCatalog = (data) => {
  return async (dispatch, getState) => {
    dispatch(creatingCatalog());

    const response = await axios({
      url: baseUrl,
      method: 'post',
      data: data,
    }).catch((error) => {
      dispatch(createCatalogFailure(error.message));
    });

    if (response) {
      dispatch(createCatalogSuccess(response.data));
    }
  };
};

export const updateCatalog = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}`;

    dispatch(updatingCatalog());

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    }).catch((error) => {
      dispatch(updateCatalogFailure(error.message));
    });

    if (response) {
      dispatch(updateCatalogSuccess(response.data));
    }
  };
};

export const deleteCatalog = (id) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}`;

    dispatch(deletingCatalog());

    const response = await axios({
      url: url,
      method: 'delete',
    }).catch((error) => {
      dispatch(deleteCatalogFailure(error.message));
    });

    if (response) {
      dispatch(deleteCatalogSuccess(id));
    }
  };
};

export const getCatalog = (id) => {
  return async (dispatch, getState) => {
    const {
      catalogs: { items },
    } = getState();

    if (items[id]) {
      const catalog = { ...items[id] };
      dispatch(getCatalogSuccess(catalog));
      return;
    }

    dispatch(loadingCatalogs());

    const response = await axios({
      url: `${baseUrl}/${id}`,
      method: 'get',
    }).catch((error) => {
      dispatch(getCatalogFailure(error.message));
    });

    if (response) {
      const catalog = response.data;
      catalog.products = getIds(catalog.products || []);
      dispatch(getCatalogSuccess(catalog));
    }
  };
};

const loadingCatalogs = () => {
  return {
    type: LOADING_CATALOGS,
  };
};

const setCatalogsListTotal = (total) => {
  return {
    type: SET_CATALOGS_LIST_TOTAL,
    payload: total,
  };
};

const addListRequest = (req) => {
  return {
    type: ADD_CATALOGS_LIST_REQUEST,
    payload: req,
  };
};

const setListCurrentPage = (ids) => {
  return {
    type: SET_CATALOGS_LIST_CURRENT_PAGE,
    payload: ids,
  };
};

const getCatalogSuccess = (catalog) => {
  return {
    type: GET_CATALOG_SUCCESS,
    payload: catalog,
  };
};

const getCatalogFailure = (message) => {
  return {
    type: GET_CATALOG_FAILURE,
    payload: message,
  };
};

export const loadCatalogsSuccess = (catalogs) => {
  return {
    type: LOAD_CATALOGS_SUCCESS,
    payload: {
      items: buildObjectOfItems(catalogs),
    },
  };
};

const loadCatalogsFailure = (message) => {
  return {
    type: LOAD_CATALOGS_FAILURE,
    payload: message,
  };
};

const creatingCatalog = () => {
  return {
    type: CREATING_CATALOG,
  };
};

const createCatalogSuccess = (catalog) => {
  return {
    type: CREATE_CATALOG_SUCCESS,
    payload: catalog,
  };
};

const createCatalogFailure = (message) => {
  return {
    type: CREATE_CATALOG_FAILURE,
    payload: message,
  };
};

const updatingCatalog = () => {
  return {
    type: UPDATING_CATALOG,
  };
};

const updateCatalogSuccess = (catalog) => {
  return {
    type: UPDATE_CATALOG_SUCCESS,
    payload: catalog,
  };
};

const updateCatalogFailure = (message) => {
  return {
    type: UPDATE_CATALOG_FAILURE,
    payload: message,
  };
};

const deletingCatalog = () => {
  return {
    type: DELETING_CATALOG,
  };
};

const deleteCatalogSuccess = (id) => {
  return {
    type: DELETE_CATALOG_SUCCESS,
    payload: id,
  };
};

const deleteCatalogFailure = (message) => {
  return {
    type: DELETE_CATALOG_FAILURE,
    payload: message,
  };
};
