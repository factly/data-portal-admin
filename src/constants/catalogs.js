export const LOADING_CATALOGS = 'LOADING_CATALOGS';
export const LOAD_CATALOGS_SUCCESS = 'LOAD_CATALOGS_SUCCESS';
export const SET_CATALOGS_LIST_TOTAL = 'SET_CATALOGS_LIST_TOTAL';
export const ADD_CATALOGS_LIST_REQUEST = 'ADD_CATALOGS_LIST_REQUEST';
export const SET_CATALOGS_LIST_CURRENT_PAGE = 'SET_CATALOGS_LIST_CURRENT_PAGE';
export const LOAD_CATALOGS_FAILURE = 'LOAD_CATALOGS_FAILURE';
export const CREATING_CATALOG = 'CREATING_CATALOG';
export const CREATE_CATALOG_SUCCESS = 'CREATE_CATALOG_SUCCESS';
export const CREATE_CATALOG_FAILURE = 'CREATE_CATALOG_FAILURE';
export const UPDATING_CATALOG = 'UPDATING_CATALOG';
export const UPDATE_CATALOG_SUCCESS = 'UPDATE_CATALOG_SUCCESS';
export const UPDATE_CATALOG_FAILURE = 'UPDATE_CATALOG_FAILURE';
export const DELETING_CATALOG = 'DELETING_CATALOG';
export const DELETE_CATALOG_SUCCESS = 'DELETE_CATALOG_SUCCESS';
export const DELETE_CATALOG_FAILURE = 'DELETE_CATALOG_FAILURE';

export const baseUrl = `${process.env.REACT_APP_API_URL}/catalogs`;
