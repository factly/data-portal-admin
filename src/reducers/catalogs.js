import {
  ADD_CATALOGS_LIST_REQUEST,
  SET_CATALOGS_LIST_CURRENT_PAGE,
  LOADING_CATALOGS,
  LOAD_CATALOGS_SUCCESS,
  SET_CATALOGS_LIST_TOTAL,
  LOAD_CATALOGS_FAILURE,
  CREATING_CATALOG,
  CREATE_CATALOG_SUCCESS,
  CREATE_CATALOG_FAILURE,
  UPDATE_CATALOG_SUCCESS,
  UPDATE_CATALOG_FAILURE,
  DELETE_CATALOG_SUCCESS,
  DELETE_CATALOG_FAILURE,
} from '../constants/catalogs';

const initialState = {
  loading: false,
  ids: [],
  req: [],
  items: {},
  total: 0,
};

export default function catalogsReducer(state = initialState, action = {}) {
  switch (action.type) {
    case LOADING_CATALOGS:
      return {
        ...state,
        loading: true,
      };
    case LOAD_CATALOGS_SUCCESS: {
      const { items } = action.payload;
      return {
        ...state,
        loading: false,
        items: { ...state.items, ...items },
      };
    }
    case SET_CATALOGS_LIST_TOTAL:
      return {
        ...state,
        total: action.payload,
      };
    case ADD_CATALOGS_LIST_REQUEST: {
      return {
        ...state,
        req: [...state.req, action.payload],
      };
    }
    case SET_CATALOGS_LIST_CURRENT_PAGE:
      return {
        ...state,
        ids: action.payload,
      };
    case LOAD_CATALOGS_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case CREATING_CATALOG:
      return {
        ...state,
        loading: true,
      };
    case CREATE_CATALOG_SUCCESS: {
      const catalog = action.payload;
      return {
        ...state,
        loading: false,
        req: [],
        items: { ...state.items, [catalog.id]: catalog },
        total: state.total + 1,
      };
    }
    case CREATE_CATALOG_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case UPDATE_CATALOG_SUCCESS: {
      const catalog = action.payload;
      return {
        ...state,
        loading: false,
        items: { ...state.items, [catalog.id]: catalog },
      };
    }
    case UPDATE_CATALOG_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case DELETE_CATALOG_SUCCESS: {
      const id = action.payload;
      const newItems = { ...state.items };
      delete newItems[id];
      return {
        ...state,
        loading: false,
        req: [],
        ids: [],
        items: newItems,
        total: state.total - 1,
      };
    }
    case DELETE_CATALOG_FAILURE:
      return {
        ...state,
        loading: false,
      };
    default:
      return state;
  }
}
