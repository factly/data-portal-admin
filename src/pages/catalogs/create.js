import React from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import CatalogCreateForm from './components/CatalogCreateForm';
import { createCatalog } from '../../actions/catalogs';

const CatalogCreate = () => {
  const history = useHistory();
  const dispatch = useDispatch();

  const onCreate = async (values) => {
    await dispatch(createCatalog(values));
    history.push('/categories');
  };

  return <CatalogCreateForm onSubmit={onCreate} />;
};

export default CatalogCreate;
