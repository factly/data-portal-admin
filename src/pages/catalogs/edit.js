import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';

import CatalogCreate from './components/CatalogCreateForm';
import { getCatalog, updateCatalog } from '../../actions/catalogs';

const EditCatalog = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const { loading, catalog } = useSelector(({ catalogs }) => ({
    loading: catalogs.loading,
    catalog: catalogs.items[id],
  }));

  React.useEffect(() => {
    dispatch(getCatalog(id));
  }, []);

  if (loading) return <Skeleton />;

  const onUpdate = async (values) => {
    await dispatch(updateCatalog(id, values));
    history.push('/catalogs');
  };

  return <CatalogCreate data={catalog} onSubmit={onUpdate} />;
};

export default EditCatalog;
