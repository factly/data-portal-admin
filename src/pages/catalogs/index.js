import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Table, Input, Button, Popconfirm, Form, notification } from 'antd';
import moment from 'moment';
import { Link } from 'react-router-dom';
import { EditOutlined, DeleteOutlined, SaveOutlined, CloseOutlined } from '@ant-design/icons';

import { loadCatalogs, updateCatalog, deleteCatalog } from '../../actions/catalogs';

const EditableCell = ({ editing, dataIndex, title, record, index, children, ...restProps }) => {
  return (
    <td {...restProps}>
      {editing ? (
        <Form.Item
          name={dataIndex}
          style={{
            margin: 0,
          }}
          rules={[
            {
              required: true,
              message: `Please enter ${title}!`,
            },
          ]}
        >
          <Input />
        </Form.Item>
      ) : (
        children
      )}
    </td>
  );
};

const Catalogs = (props) => {
  const [form] = Form.useForm();
  const [editingKey, setEditingKey] = useState('');
  const { data, total, load, update, remove } = props;
  const [pagination, setPagination] = useState({
    current: 1,
    defaultPageSize: 5,
    pageSize: 5,
    total,
  });

  React.useEffect(() => {
    handleTableChange(pagination);
  }, [total]);

  const handleTableChange = ({ current, pageSize }) => {
    cancel();
    load(current, pageSize);
    setPagination({ ...pagination, current, pageSize, total });
  };

  const isEditing = (record) => record.id === editingKey;

  const edit = (record) => {
    form.setFieldsValue(record);
    setEditingKey(record.id);
  };

  const cancel = () => {
    setEditingKey('');
  };

  const save = async (key) => {
    try {
      const row = await form.validateFields();
      const index = data.findIndex((item) => item.id === key);

      if (index > -1) {
        update(key, row)
          .then(() => {
            setEditingKey('');
            notification.success({
              message: 'Success',
              description: 'Catalog succesfully updated',
            });
          })
          .catch(() => {
            notification.error({
              message: 'Error',
              description: 'Something went wrong',
            });
          });
      }
    } catch (err) {
      notification.warning({
        message: 'Warning',
        description: 'Validation failed',
      });
    }
  };

  const deleteCatalog = (key) => {
    const index = data.findIndex((item) => item.id === key);
    if (index > -1) {
      remove(key)
        .then(() => {
          notification.success({
            message: 'Success',
            description: 'Catalog succesfully deleted',
          });
        })
        .catch(() => {
          notification.error({
            message: 'Error',
            description: 'Something went wrong',
          });
        });
    }
  };

  const columns = [
    {
      title: 'Title',
      dataIndex: 'title',
      width: '20%',
      editable: true,
    },
    {
      title: 'Description',
      dataIndex: 'description',
      width: '20%',
      editable: true,
    },
    {
      title: 'Price',
      dataIndex: 'price',
      width: '20%',
      editable: true,
    },
    {
      title: 'Published At',
      dataIndex: 'published_date',
      width: '20%',
      render: (_, record) => {
        return <span title={record.created_at}>{moment(record.created_at).fromNow()}</span>;
      },
    },
    {
      title: 'Operation',
      dataIndex: 'operation',
      render: (_, record) => {
        const editable = isEditing(record);
        return editable ? (
          <span>
            <Button
              type="primary"
              icon={<SaveOutlined />}
              onClick={() => save(record.id)}
              style={{
                marginRight: 8,
              }}
            >
              Save
            </Button>
            <Popconfirm title="Sure to cancel?" onConfirm={cancel}>
              <Button icon={<CloseOutlined />}>Cancel</Button>
            </Popconfirm>
          </span>
        ) : (
          <span>
            <Button
              type="primary"
              icon={<EditOutlined />}
              disabled={editingKey !== ''}
              onClick={() => edit(record)}
              style={{
                marginRight: 8,
              }}
            >
              Edit
            </Button>
            <Popconfirm
              disabled={editingKey !== ''}
              title="Sure to delete?"
              onConfirm={() => deleteCatalog(record.id)}
            >
              <Button disabled={editingKey !== ''} icon={<DeleteOutlined />}>
                Delete
              </Button>
            </Popconfirm>
          </span>
        );
      },
    },
  ];

  const mergedColumns = columns.map((col) => {
    if (!col.editable) {
      return col;
    }

    return {
      ...col,
      onCell: (record) => ({
        record,
        dataIndex: col.dataIndex,
        title: col.title,
        editing: isEditing(record),
      }),
    };
  });

  return (
    <div>
      <Link to={'/catalogs/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Catalog
        </Button>
      </Link>
      <Form form={form} component={false}>
        <Table
          components={{
            body: {
              cell: EditableCell,
            },
          }}
          bordered
          rowKey="id"
          dataSource={data}
          onChange={handleTableChange}
          columns={mergedColumns}
          rowClassName="editable-row"
          pagination={pagination}
        />
      </Form>
    </div>
  );
};

Catalogs.propTypes = {
  data: PropTypes.array.isRequired,
  total: PropTypes.number.isRequired,
  load: PropTypes.func.isRequired,
  update: PropTypes.func.isRequired,
  remove: PropTypes.func.isRequired,
};

const mapStateToProps = ({ catalogs }) => {
  const { ids, items, total } = catalogs;
  return {
    data: ids.map((id) => items[id]),
    total,
  };
};

const mapDispatchToProps = (dispatch) => ({
  load: (page, limit) => dispatch(loadCatalogs(page, limit)),
  update: (id, data) => dispatch(updateCatalog(id, data)),
  remove: (id) => dispatch(deleteCatalog(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Catalogs);
