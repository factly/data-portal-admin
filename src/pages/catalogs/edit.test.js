import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider, useSelector } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import EditCatalog from './edit';
import * as actions from '../../actions/catalogs';
import CatalogEditForm from './components/CatalogCreateForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
  useParams: jest.fn().mockReturnValue({ id: '1' }),
}));

jest.mock('../../actions/catalogs', () => ({
  getCatalog: jest.fn(),
  updateCatalog: jest.fn(),
}));

describe('Catalogs List component', () => {
  const store = mockStore({
    catalogs: {
      loading: false,
      ids: [],
      req: [],
      items: {
        1: {
          id: 1,
          title: 'Catalog-1',
          description: 'description',
          price: 100,
          published_date: null,
        },
        2: {
          id: 2,
          title: 'Catalog-2',
          description: 'description',
          price: 100,
          published_date: null,
        },
      },
      total: 0,
    },
    products: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
  });
  store.dispatch = jest.fn(() => ({}));

  const mockedDispatch = jest.fn();
  useDispatch.mockReturnValue(mockedDispatch);

  describe('snapshot testing', () => {
    it('should render the component', () => {
      useSelector
        .mockReturnValueOnce({
          catalog: {
            id: 1,
            title: 'Catalog-1',
            description: 'description',
            price: 100,
            published_date: null,
          },
          loading: false,
        })
        .mockReturnValueOnce({ details: [], total: 0, loading: false });

      let component;
      act(() => {
        component = shallow(
          <Provider store={store}>
            <EditCatalog />
          </Provider>,
        );
      });
      // const tree = component.toJSON();
      expect(component).toMatchSnapshot();
    });
    it('should match component with empty data', () => {
      useSelector
        .mockReturnValueOnce({
          catalog: {},
          loading: false,
        })
        .mockReturnValueOnce({ details: [], total: 0, loading: false });
      let component;
      store.details = {};
      act(() => {
        component = shallow(
          <Provider store={store}>
            <EditCatalog />
          </Provider>,
        );
      });
      // const tree = component.toJSON();
      expect(component).toMatchSnapshot();
    });
    it('should match skeleton while loading', () => {
      useSelector
        .mockReturnValueOnce({
          catalog: {},
          loading: true,
        })
        .mockReturnValueOnce({ details: [], total: 0, loading: false });
      let component;
      act(() => {
        component = shallow(
          <Provider store={store}>
            <EditCatalog />
          </Provider>,
        );
      });
      // const tree = component.toJSON();
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call get action', () => {
      useSelector.mockReset();
      useSelector
        .mockReturnValueOnce({ catalog: null, loading: true })
        .mockReturnValueOnce({ details: [], total: 0, loading: false });
      actions.getCatalog.mockReset();
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditCatalog />
          </Provider>,
        );
      });
      expect(actions.getCatalog).toHaveBeenCalledWith('1');
    });
    it('should call updateCatalog', (done) => {
      useSelector.mockReset();
      useSelector
        .mockReturnValueOnce({ catalog: {}, loading: false })
        .mockReturnValueOnce({ details: [], total: 0, loading: false });
      actions.updateCatalog.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditCatalog />
          </Provider>,
        );
      });
      wrapper.find(CatalogEditForm).props().onSubmit({ test: 'test' });
      setTimeout(() => {
        expect(actions.updateCatalog).toHaveBeenCalledWith('1', { test: 'test' });
        expect(push).toHaveBeenCalledWith('/catalogs');
        done();
      }, 0);
    });
  });
});
