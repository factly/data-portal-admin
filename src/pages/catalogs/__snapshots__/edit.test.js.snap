// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalogs List component snapshot testing should match component with empty data 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "details": Object {},
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "details": Object {},
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <EditCatalog />
</ContextProvider>
`;

exports[`Catalogs List component snapshot testing should match skeleton while loading 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "details": Object {},
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "details": Object {},
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <EditCatalog />
</ContextProvider>
`;

exports[`Catalogs List component snapshot testing should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <EditCatalog />
</ContextProvider>
`;
